worker/firewaller/firewaller.go

```diff
@@ -1296,17 +1314,59 @@ func (rd *remoteRelationData) watchLoop() error {
 		}
 	}
 
+	if rd.isOffer {
+		// On the offering side we watch the local model for ingress changes
+		// which will have been published from the consuming model.
+		rd.ingressAddressWatcherFunc = func() (watcher.StringsWatcher, error) {
+			return rd.fw.firewallerApi.WatchIngressAddressesForRelation(rd.tag)
+		}
+	} else {
+		if rd.endpointRole == charm.RoleProvider {
+			// On the consuming side, if this is the provider end of the relation,
+			// we watch the remote model's egress changes to get our ingress changes.
+			apiInfo, err := rd.fw.firewallerApi.ControllerAPIInfoForModel(rd.remoteModelUUID)
+			if err != nil {
+				return errors.Annotatef(err, "cannot get api info for model %v", rd.remoteModelUUID)
+			}
+			rd.crossModelFirewallerFacade, err = rd.fw.newRemoteFirewallerAPIFunc(apiInfo)
+			if err != nil {
+				return errors.Annotate(err, "cannot open facade to remote model to publish network change")
+			}
+
+			rd.ingressAddressWatcherFunc = func() (watcher.StringsWatcher, error) {
+				mac, err := rd.fw.firewallerApi.MacaroonForRelation(rd.tag.Id())
+				if err != nil {
+					return nil, errors.Annotatef(err, "cannot get macaroon for %v", rd.tag.Id())
+				}
+				arg := params.RemoteRelationArg{
+					RemoteEntityId: *rd.remoteRelationId,
+					Macaroons:      macaroon.Slice{mac},
+				}
+				return rd.crossModelFirewallerFacade.WatchEgressAddressesForRelation(arg)
+			}
+		}
+		rd.egressAddressWatcherFunc = rd.fw.firewallerApi.WatchEgressAddressesForRelation
```

Comment:
If we're on a consuming side with a provider role, we have both ingress and egress address watcher funcs set up. But the provider endpoint loop doesn't use the egress watcher func, and trying to think it through I don't see how it could - should this just be in the else part for the preceding if?

I guess this is academic if you just make the func fields methods instead.